lib/LongVectorLoweringPass.cpp

```diff
@@ -34,16 +44,377 @@ class LongVectorLoweringPass final : public ModulePass {
+    }
+  }
+
+  return NewValue;
+}
+
+using ScalarOperationFactory =
+    std::function<Value *(IRBuilder<> & /* B */, ArrayRef<Value *> /* Args */)>;
+
+/// Scalarise a vector instruction element-wise by invoking the operation
+/// @p ScalarOperation.
+Value *convertVectorOperation(IRBuilder<> &B, Type *EquivalentReturnTy,
+                              ArrayRef<Value *> EquivalentArgs,
+                              ScalarOperationFactory ScalarOperation) {
+  assert(EquivalentReturnTy != nullptr);
+  assert(EquivalentReturnTy->isStructTy());
+
+  Value *ReturnValue = UndefValue::get(EquivalentReturnTy);
+  unsigned Arity = EquivalentReturnTy->getNumContainedTypes();
+
+  // Invoke the scalar operation once for each vector element.
+  for (unsigned i = 0; i < Arity; ++i) {
+    SmallVector<Value *, 16> Args;
+    Args.resize(EquivalentArgs.size());
+
+    for (unsigned j = 0; j < Args.size(); ++j) {
+      assert(EquivalentArgs[j]->getType()->isStructTy());
+      Args[j] = B.CreateExtractValue(EquivalentArgs[j], i);
+    }
+
+    Value *Scalar = ScalarOperation(B, Args);
+    ReturnValue = B.CreateInsertValue(ReturnValue, Scalar, i);
+  }
+
+  return ReturnValue;
+}
+
+bool LongVectorLoweringPass::runOnModule(Module &M) {
   bool Modified = false;
+
+  for (auto &F : M.functions()) {
+    Modified |= runOnFunction(F);
+  }
+
+  return Modified;
+}
+
+Value *LongVectorLoweringPass::visit(Value *V) {
+  // Already handled?
+  auto it = InstructionMap.find(V);
+  if (it != InstructionMap.end()) {
+    return it->second;
+  }
+
+  if (auto *BinOp = dyn_cast<BinaryOperator>(V)) {
+    if (auto *EquivalentType = getEquivalentType(BinOp->getType())) {
+      IRBuilder<> B(BinOp);
+
+      SmallVector<Value *, 16> EquivalentArgs;
+      for (auto &Operand : BinOp->operands()) {
+        auto *Arg = Operand.get();
+        // TODO Visit argument to lower it.
+        // Instead, for now, we create an equivalent aggregate.
+        auto *EquivalentArgTy = getEquivalentType(Arg->getType());
+        auto *EquivalentArg = convertEquivalentValue(B, Arg, EquivalentArgTy);
+        EquivalentArgs.push_back(EquivalentArg);
+      }
+
+      unsigned Opcode = BinOp->getOpcode();
+      Value *EquivalentValue = convertVectorOperation( //
+          B, EquivalentType, EquivalentArgs,           //
+          [Opcode](auto &B, auto Args) {
+            return B.CreateNAryOp(Opcode, Args);
+          });
+
+      // TODO Implement support for additional instructions.
+      // Because support is very limited, as an initial step we convert the
+      // aggregate back to a vector to generate a valid module.
+      auto *ReplacementValue =
+          convertEquivalentValue(B, EquivalentValue, BinOp->getType());
+      registerReplacement(*BinOp, *ReplacementValue);
+
+      return ReplacementValue;
+    }
+  }
+
+#ifndef NDEBUG
+  dbgs() << "Value not handled: " << *V << '\n';
+#endif
+  // TODO Once additional features are implemented, turn this into an error by
```

Comment:
The default case is indeed that no lowering is required. In a latter commit, when using `InstVisitor`, this will be handled more elegantly and, of course, won't generate an error.

This commented `llvm_unreachable` is meant to catch cases that are not implemented yet. I do not expect cases that cannot be lowered so this should be seen as a bug in the implementation.